.ci

```diff
@@ -0,0 +1,158 @@
+def k8s_label = "${UUID.randomUUID().toString()}"
+
+pipeline {
+  environment {
+    APP_NAME = 'bulk-processor'
+    APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE = 'true'
+    CI = 'true'
+    CLOUDSDK_CORE_DISABLE_PROMPTS = '1'
+    GIT_COMMIT_SHORT = sh(script: "git rev-parse --short ${GIT_COMMIT}", returnStdout: true).trim()
+    GIT_MESSAGE = sh(script: "git log --format=%B -n 1 ${GIT_COMMIT}", returnStdout: true).trim()
+    GIT_USER = sh(script: "git log -1 --pretty=format:'%ae'", returnStdout: true).trim()
+    GITHUB_URL = "https://github.com"
+    LANG = "en_US.UTF-8"
+    LANGUAGE = "en_US:en"
+    LC_ALL = "en_US.UTF-8"
+    PRODUCTION_DEPLOY="false"
+    SLACK_CHANNEL = "#staging-releases"
+  } // environment
+
+  agent {
+    kubernetes {
+      label k8s_label
+      defaultContainer 'jnlp'
+      yaml """
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test
+spec:
+  restartPolicy: Never
+  containers:
+  - name: ruby
+    image: gcr.io/alist-development/ruby:2.7.5
+    imagePullPolicy: Always
+    resources:
+      requests:
+        memory: "1024Mi"
+        cpu: "1"
+      requests:
+        memory: "1024Mi"
+        cpu: "1"
+    command:
+    - "tail"
+    - "-f"
+    - "/dev/null"
+"""
+    } // kubernetes
+  } // agent
+
+  options {
+    timestamps()
+    timeout(time: 10, unit: 'MINUTES')
+    ansiColor('xterm')
+  } // options
+
+  stages {
+    stage('Preparation') {
+      parallel {
+        stage('Slack') {
+          steps {
+            slackPreparation()
+          } // steps
+        } // stage - Slack
+        stage('Build Description') {
+          steps {
+            buildDescription()
+          }
+        }
+        stage('Bundle') {
+          steps {
+            withCredentials([usernamePassword(credentialsId: 'ApartmentList-CI', usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_TOKEN')]) {
+              container('ruby') {
+                withEnv([
+                  "GIT_USERNAME=${GIT_USERNAME}",
+                  "GIT_TOKEN=${GIT_TOKEN}"
+                ]) {
+                  sh 'apt-get update'
+                  sh 'gem install bundler:2.2.6 --no-post-install-message --no-document'
+                  sh 'bundle config set --global rubygems.pkg.github.com $GIT_USERNAME:$GIT_TOKEN'
+                  sh 'bundle install -j 12'
+                } // withEnv
+              } //container
+            } // withCredential
+          } //steps
+        } // stage - Bundle
+      } //parallel
+    } //stage - Preparation
+
+    stage('Testing') {
+      when {
+        allOf {
+          changeRequest target: 'main'
+        }
+      }
+
+      parallel {
+        stage('Run RSpec') {
+          steps {
+            container('ruby') {
+              withEnv([
+                "BUNDLE_GITHUB__COM=${BUNDLE_GITHUB__COM}",
+                "BUNDLE_GITHUB__HTTPS=${BUNDLE_GITHUB__HTTPS}"
+              ]) {
+                sh label: 'RSpec', script: 'bundle exec rake'
+              } // withEnv
```

Comment:
I think you can pull `sh()` out of `withEnv()` block because you now use package manager, not github code.